### PR TITLE
Df apply generic types

### PR DESCRIPTION
**Pull request by jbrezmorf**
Apply generic types to Fields.

### CODE REVIEW

File: src/fields/field.impl.hh [hunk 1]
```diff
@@ -436,8 +405,8 @@ void Field<spacedim,Value>::check_initialized_region_fields_() {
     	std::string region_list;
     	// has to deal with fact that reader can not deal with input consisting of simple values
     	string default_input=input_default();
-    	auto input_type = get_input_type();
-        Input::ReaderToStorage reader( default_input, input_type, Input::FileFormat::format_JSON );
+    	auto input_type = get_input_type().make_instance().first;
```
Comment (jbrezmorf) [inline src/fields/field.impl.hh]: Repeated call to make_instance() is not explicitly guaranteed to return exactly the same type. Remember types produced by instances and return them on second call.

File: src/fields/field_algo_base.hh [hunk 2]
```diff
@@ -75,7 +76,13 @@ public:
...
         * given by element_input_type pointer.
         */
-       static Input::Type::AbstractRecord & get_input_type(const typename Value::ElementInputType *element_input_type=nullptr);
+       static Input::Type::AbstractRecord & get_input_type();
+
+       /**
+        * Returns parameterized whole tree of input types for FieldBase with all descendants based on element input type (namely
+        * for FieldConstant) given by element_input_type pointer.
+        */
+       static const Input::Type::Instance & get_input_type_instance(const Input::Type::Selection *value_selection=NULL);
```
Comment (jbrezmorf) [inline src/fields/field_algo_base.hh]: Is this still necessary?

File: src/fields/field_algo_base.impl.hh [hunk 3]
```diff
@@ -53,31 +54,24 @@ string FieldAlgorithmBase<spacedim, Value>::template_name() {
...
 
 template <int spacedim, class Value>
-Input::Type::AbstractRecord & FieldAlgorithmBase<spacedim, Value>::get_input_type(const typename Value::ElementInputType *element_input_type) {
-	const it::Selection *element_input_sel = nullptr;
-	if (element_input_type != nullptr) {
-		const it::Selection * sel_type = dynamic_cast<const it::Selection *>(element_input_type);
-		if (sel_type != NULL ) element_input_sel = sel_type;
-	}
+Input::Type::AbstractRecord & FieldAlgorithmBase<spacedim, Value>::get_input_type() {
+	return it::AbstractRecord("Field:"+template_name(), "Abstract record for all time-space functions.")
```
Comment (jbrezmorf) [inline src/fields/field_algo_base.impl.hh]: - AbstractRecord::content_hash() should has also attribute "parameters". Then we can set template_name() to contain just specification of dimension.

File: src/fields/field_algo_base.impl.hh [hunk 4]
```diff
@@ -53,31 +54,24 @@ string FieldAlgorithmBase<spacedim, Value>::template_name() {
...
-#endif
-		type.add_child( const_cast<it::Record &>(FieldInterpolatedP0<spacedim,Value>::get_input_type(type, element_input_type)) );
-		type.add_child( const_cast<it::Record &>(FieldElementwise<spacedim,Value>::get_input_type(type, element_input_type)) );
-
-		type.finish();
-    }
 
-    return type.close();
+template <int spacedim, class Value>
+const Input::Type::Instance & FieldAlgorithmBase<spacedim, Value>::get_input_type_instance(const Input::Type::Selection *value_selection) {
```
Comment (jbrezmorf) [inline src/fields/field_algo_base.impl.hh]: Can remove?

File: src/fields/field_algo_base.impl.hh [hunk 5]
```diff
@@ -53,31 +54,24 @@ string FieldAlgorithmBase<spacedim, Value>::template_name() {
...
+	std::vector<it::TypeBase::ParameterPair> param_vec;
+	if ( boost::is_same<typename Value::element_type, FieldEnum>::value && value_selection) {
+		param_vec.push_back( std::make_pair("element_input_type", boost::make_shared<it::Selection>(*value_selection)) );
```
Comment (jbrezmorf) [inline src/fields/field_algo_base.impl.hh]: ASSERT(value_selection, "..");

File: src/fields/field_algo_base.impl.hh [hunk 6]
```diff
@@ -53,31 +54,24 @@ string FieldAlgorithmBase<spacedim, Value>::template_name() {
...
-		type.add_child( const_cast<it::Record &>(FieldElementwise<spacedim,Value>::get_input_type(type, element_input_type)) );
-
-		type.finish();
-    }
 
-    return type.close();
+template <int spacedim, class Value>
+const Input::Type::Instance & FieldAlgorithmBase<spacedim, Value>::get_input_type_instance(const Input::Type::Selection *value_selection) {
+	std::vector<it::TypeBase::ParameterPair> param_vec;
+	if ( boost::is_same<typename Value::element_type, FieldEnum>::value && value_selection) {
```
Comment (jbrezmorf) [inline src/fields/field_algo_base.impl.hh]: Remove &&value_selection

File: src/fields/field.impl.hh [hunk 7]
```diff
@@ -98,11 +98,11 @@ Field<spacedim,Value> &Field<spacedim,Value>::operator=(const Field<spacedim,Val
 
 
 template<int spacedim, class Value>
-it::AbstractRecord &Field<spacedim,Value>::get_input_type() {
+const it::Instance &Field<spacedim,Value>::get_input_type() {
 	if (is_enum_valued) {
```
Comment (jbrezmorf) [inline src/fields/field.impl.hh]: Remove test.   Move definition of is_enum_valued into FieldAlgoBase.

File: src/input/type_base.cc [hunk 8]
```diff
@@ -238,6 +242,12 @@ Array Array::deep_copy() const {
 }
 
 
+Array &Array::root_of_generic_subtree() {
```
Comment (jbrezmorf) [inline src/input/type_base.cc]: Remove.

File: src/input/type_base.hh [hunk 9]
```diff
@@ -524,6 +535,9 @@ public:
     }
 
 
+    virtual MakeInstanceReturnType make_instance(std::vector<ParameterPair> vec = std::vector<ParameterPair>()) const override;
```
Comment (jbrezmorf) [inline src/input/type_base.hh]: Not virtual. Check also other override methods.

File: src/input/type_record.cc [hunk 10]
```diff
@@ -240,7 +240,16 @@ bool Record::finish(bool is_generic)
     {
     	if (it->key_ != "TYPE") {
 			if (typeid( *(it->type_.get()) ) == typeid(Instance)) it->type_ = it->type_->make_instance().first;
-            data_->finished = data_->finished && it->type_->finish(is_generic);
+            try {
+            	data_->finished = data_->finished && it->type_->finish(is_generic);
+            } catch (ExcParamaterInIst &e) {
+            	if (root_of_generic_subtree_) {
```
Comment (jbrezmorf) [inline src/input/type_record.cc]: See new comment in issue #415

File: src/fields/field.impl.hh (hunk 1 above)
Comment (dflanderka) [inline src/fields/field.impl.hh]: Done in Instance::make_instance

File: src/fields/field.hh [hunk 11]
```diff
@@ -55,8 +55,8 @@ public:
     typedef FieldAlgorithmBase<spacedim, Value> FieldBaseType;
     typedef std::shared_ptr< FieldBaseType > FieldBasePtr;
     typedef typename FieldAlgorithmBase<spacedim, Value>::Point Point;
+    typedef Value ValueType;
```
Comment (jbrezmorf) [inline src/fields/field.hh]: Why this typdef?

**Comment by jbrezmorf** [comment]
Possibly remove the typedef and merge yourself.

File: src/fields/field.hh (hunk 11 above)
Comment (dflanderka) [inline src/fields/field.hh]: This typedef is used, but same typedef in FieldAlgoBase is unnecessary - removed.

**Pull request merged**